input_validator.rb

```diff
@@ -0,0 +1,40 @@
+require 'csv'
+require_relative 'string'
+
+class InputValidator
+  attr_reader :x_coordinate, :y_coordinate, :filename
+  def initialize(input)
+    @x_coordinate = input[0]
+    @y_coordinate = input[1]
+    @filename = input[2]
+  end
+
+  def validate
+    raise 'Invalid input: check correctness of arguments, filename or file entries' unless valid_input?
+  end
+
+  private
+
+  def valid_input?
+    valid_file? && valid_user_coordinates? && valid_file_entries?
+  end
+
+  def valid_file?
+    CSV.read(filename)
+  rescue
+    false
+  end
+
+  def valid_user_coordinates?
+    x_coordinate.numeric? && y_coordinate.numeric?
+  end
+
+  def valid_file_entries?
+    CSV.read(filename).each do |entry|
```

Comment:
There are 3 CSV.read in different places. The implementation is highly coupled and very hard to change / extend. Also reading a file 3 times is not so efficient.